### PR TITLE
Translator augment

### PR DESCRIPTION
**Pull request by SidVeld**
Adds an augmented translator to the game, with which the user can understand all human languages. I did not add a translator from xeno languages, as it would have been very strong, given the difficulty of translation and the peculiarities of their languages.

I have decomposed the original augment file to improve readability and code handling. Hopefully this won't cause trouble in the future. (Sorry, but an 800+ line file is a very large file.)

### CODE REVIEW

**Comment by DreamySkrell** [comment]
not sure if adding it to the loadout is wanted
but the code refactor and splitting up files is nice

and if not wanted in the loadout, the augment could still be added, just not for crew to use

**Comment by SidVeld** [comment]
!review

**Comment by Sniblet** [comment]
Elyrans forgotten as always
Everybody's opposed to opening up languages at all for reasons that I can't explain. One-way translators would make a world of sense given the uses for it and the fact that the technology literally, mechanically already exists (integrated electronics) and also mostly exists in real life, and I'm personally in a constant hostile relationship with the language limits on IPCs especially, but I don't expect this to go through.
The fact that it's so easy to get, and so justifiable in a setting like Horizon, means probably everyone without a religious exemption is going to want to have one. And then what's the point of having separate languages? Fluff? Fluff in my flipping roleplaying game?

But anyway, if this is to be implemented, Elyran Standard is a human language. It's reasonable for a translation augment to not be packaged with Elyran, but then you just can't say that it has all human languages, because it doesn't.

**Comment by SidVeld** [comment]
OH, CRAP, I FORGOT ABOUT ELYRANS!


[AALIYAH_WROTE_JAY_AN_APOLOGY_LETTER📃__MARK_WANT_DESIREE_AND_HANNAH!😱_4-32_screenshot_(1)-1515311997.jpg](https://github.com/Aurorastation/Aurora.3/assets/22749671/748c343e-45ec-46c9-920f-1b3c9a2fa7e4)

File: code/modules/mob/say.dm [hunk 1]
```diff
@@ -76,11 +75,21 @@
...
 		return TRUE
 
-	//Language check.
+	// Language check.
 	for(var/datum/language/L in src.languages)
 		if(speaking.name == L.name)
 			return TRUE
 
+	// Try to translate with augment if it is installed and not broken.
+	if(ishuman(src))
```
Comment (Geevies) [inline code/modules/mob/say.dm:84]: best practice is not to use istype src checks, instead you should do something like this:

```
/mob/living/carbon/human/say_understands(var/mob/other, var/datum/language/speaking = null)
	. = ..()
	if(!.)
		your translator code here
```

File: code/modules/mob/say.dm [hunk 2]
```diff
@@ -76,11 +75,21 @@
...
+		var/mob/living/carbon/human/human = src
+		var/obj/item/organ/internal/augment/translator/translator = human.internal_organs_by_name[BP_AUG_TRANSLATOR]
+
+		if(translator && !translator.is_broken())
+			for(var/language_name in translator.languages)
+				if(speaking.name == language_name)
+					return TRUE
```
Comment (Geevies) [inline code/modules/mob/say.dm:91]: i'm not 100% sure, but i think you can simplify this bit by replacing the for loop and the if with this:
`if(speaking.name in translator.languages)`

File: code/modules/organs/subtypes/augment/augments/timepeace.dm [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/obj/item/organ/internal/augment/timepiece
```
Comment (Geevies) [inline code/modules/organs/subtypes/augment/augments/timepeace.dm:1]: rename the file to `timepiece.dm`

File: code/modules/organs/subtypes/augment/augments/zengu_plate.dm [hunk 4]
```diff
@@ -0,0 +1,9 @@
+/obj/item/organ/internal/augment/zenghu_plate
+	name = "zeng-hu veterancy plate "
```
Comment (Geevies) [inline code/modules/organs/subtypes/augment/augments/zengu_plate.dm:2]:
```suggestion
	name = "zeng-hu veterancy plate"
```

File: code/modules/organs/subtypes/augment/augments/zengu_plate.dm [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+	desc = " A clear sign of Zeng-Hu's best, this plate bearing the company's symbol is installed on those who prove themselves in the hyper-competitive environment."
```
Comment (Geevies) [inline code/modules/organs/subtypes/augment/augments/zengu_plate.dm:3]:
```suggestion
	desc = "A clear sign of Zeng-Hu's best, this plate bearing the company's symbol is installed on those who prove themselves in the hyper-competitive environment."
```

File: html/changelogs/sidveld-augment-translator.yml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+author: SidVeld
+
+delete-after: True
+
+changes:
+  - rscadd: "Added an augment translator that can translate human languages."
```
Comment (Geevies) [inline html/changelogs/sidveld-augment-translator.yml:6]: i think be a little more explicit here so people don't get confused and angry. maybe something like `Added an universal translator augment that lets the user understand but not speak human languages.`

**Review by Geevies — CHANGES_REQUESTED** [review]
thanks for splitting the files, that file was getting a little big

i agree that this aug is cool, and even if we don't keep it in the loadout, having it be an antag or ghost role item would be cool

File: code/modules/mob/say.dm (hunk 2 above)
Comment (SidVeld) [reply code/modules/mob/say.dm:91]: Yes, it turns out you can do that and it's much more pleasing to the eye

**Comment by SidVeld** [comment]
@Geevies Thanks for the review. Your comments helped make some good edits.

**Review by Geevies — APPROVED** [review]
lgtm!

File: code/modules/client/preference_setup/loadout/items/augments.dm [hunk 7]
```diff
@@ -255,3 +255,10 @@
 	allowed_roles = list("Physician", "Surgeon", "Pharmacist", "First Responder", "Psychiatrist", "Medical Intern", "Medical Personnel", "Security Officer", "Warden", "Security Cadet", "Investigator", "Security Personnel", "Corporate Liaison", "Assistant", "Off-Duty Crew Member")
 	faction = "Private Military Contracting Group"
 	cost = 1
+
+/datum/gear/augment/translator
+	display_name = "universal translator"
+	description = "Augmentation that allows the user to understand other human languages, but not speak them."
+	path = /obj/item/organ/internal/augment/translator
+	cost = 1
```
Comment (NonQueueingMatt) [inline code/modules/client/preference_setup/loadout/items/augments.dm:263]: This is not wanted and goes against our current internal guidelines for languages / what we want in the future from them. Gotta remove this from the loadout, sorry.

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]

File: code/modules/client/preference_setup/loadout/items/augments.dm (hunk 7 above)
Comment (SidVeld) [reply code/modules/client/preference_setup/loadout/items/augments.dm:263]: Oh, okay. I'll do it now

**Review by NonQueueingMatt — APPROVED** [review]

**Pull request merged**